packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,231 @@
+   * unprocessed callback.
+   * Fails the test if the expected callback does not match the name of the
+   * actual callback function.
+   */
+  awaitCallback(expectedCallback: StreamStatusCallback): Promise<void> {
+    if (this.pendingStates.length > 0) {
+      expect(this.pendingStates.shift()).to.equal(expectedCallback);
+      return Promise.resolve();
+    } else {
+      const deferred = new Deferred<StreamStatusCallback>();
+      this.pendingPromises.push(deferred);
+      return deferred.promise.then(actualCallback => {
+        expect(actualCallback).to.equal(expectedCallback);
+      });
+    }
+  }
+
+  verifyNoPendingCallbacks(): void {
+    expect(this.pendingStates).to.be.empty;
+  }
+}
+
+describe('Watch Stream', () => {
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  function initializeWatchStream(): Promise<PersistentListenStream> {
+    const databaseInfo = getDefaultDatabaseInfo();
+
+    return PlatformSupport.getPlatform()
+      .loadConnection(databaseInfo)
+      .then(conn => {
+        const serializer = PlatformSupport.getPlatform().newSerializer(
+          databaseInfo.databaseId
+        );
+        return new Datastore(
+          databaseInfo,
+          queue,
+          conn,
+          new EmptyCredentialsProvider(),
+          serializer
+        );
+      })
+      .then(ds => {
+        return ds.newPersistentWatchStream(streamListener);
+      });
+  }
+
+  asyncIt('can be stopped before handshake', () => {
+    let watchStream: PersistentListenStream;
+
+    return initializeWatchStream()
+      .then(ws => {
+        watchStream = ws;
+        watchStream.start();
+        return streamListener.awaitCallback('onOpen');
+      })
+      .then(() => {
+        watchStream.stop();
+      });
+  });
+});
+
+describe('Write Stream', () => {
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  function initializeWriteStream(): Promise<PersistentWriteStream> {
+    const databaseInfo = getDefaultDatabaseInfo();
+
+    return PlatformSupport.getPlatform()
+      .loadConnection(databaseInfo)
+      .then(conn => {
+        const serializer = PlatformSupport.getPlatform().newSerializer(
+          databaseInfo.databaseId
+        );
+        return new Datastore(
+          databaseInfo,
+          queue,
+          conn,
+          new EmptyCredentialsProvider(),
+          serializer
+        );
+      })
+      .then(ds => {
+        return ds.newPersistentWriteStream(streamListener);
+      });
+  }
+
+  asyncIt('can be stopped before handshake', () => {
```

Comment:
same comment as above about not testing we don't get extra onClose event after stop().

Reply:
As above.